grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -0,0 +1,165 @@
+
+
+class JSONRequest(JSONMixin, Request):
+  pass
+
+
+class JSONResponse(JSONMixin, Response):
+  pass
+
+
+class Grrafana(object):
+  """GRRafana HTTP server instance.
+  
+  A full description of all endpoints implemented within this HTTP
+  server can be found in https://github.com/simPod/grafana-json-datasource#api."""
+
+  def __init__(self, config: dict) -> None:
+    """Constructor."""
+    self.url_map = Map([
+    Rule('/', endpoint='Root', methods=["GET"]),
+    Rule('/search', endpoint='Search', methods=["POST"]),
+    Rule('/query', endpoint='Query', methods=["POST"]),
+    Rule('/annotations', endpoint='Annotations', methods=["POST"])
+    ])
+
+  def dispatchRequest(self, request: JSONRequest) -> Any:
+    """Maps requests to different methods."""
+    adapter = self.url_map.bind_to_environ(request.environ)
+    try:
+      endpoint, values = adapter.match()
+      return getattr(self, 'on' + endpoint)(request, **values)
+    except HTTPException as e:
+      return e
+
+  def wsgiApp(self, environ, start_response) -> Response:
+    request = JSONRequest(environ)
+    response = self.dispatchRequest(request)
+    return response(environ, start_response)
+
+  def __call__(self, environ, start_response):
+    return self.wsgiApp(environ, start_response)
+
+  def onRoot(self, request: JSONRequest) -> JSONResponse:
+    """Returns OK message to database connection check."""
+    return JSONResponse()
+
+  def onSearch(self, request: JSONRequest) -> JSONResponse:
+    """Depending on the request type, returns either available client
+    resource usage metrics from Fleetspeak database, or possible values
+    for a defined Grafana variable (currently supports only variables based
+    on client IDs)."""
+    if "type" in request.json:
+      # Request issued on Panel > Queries page.
+      response = AVAILABLE_METRICS
+    else:
+      # Grafana issued request on Variables > New/Edit page.
+      response = fetchClientIds()  # todo: support Grafana variables other than ClientID.
+    return JSONResponse(response=json.dumps(response), mimetype="application/json")
+
+  def onQuery(self, request: JSONRequest) -> JSONResponse:
+    """Given a client ID as a Grafana variable and targets (resource usages),
+    returns datapoints in a format Grafana can interpret."""
+    request_json_format = request.json
+    requested_client_id = extractClientIdFromVariable(request_json_format)  # There must be a ClientID variable declated in Grafana.
+    requested_targets = [entry["target"] for entry in request_json_format["targets"]]
+    response = fetchDatapointsForTargets(requested_client_id, request_json_format["maxDataPoints"], requested_targets)
+    return JSONResponse(response=json.dumps(response), mimetype="application/json")
+
+  def onAnnotations(self, request: JSONRequest) -> JSONResponse:
+    pass
+
+
+def fetchClientIds() -> List[Text]:
+  """Fetches GRR client IDs that have resource usage records in Fleetspeak database."""
```

Comment:
I'd suggest not fetching the client ids. This will work on trivial deployments only. Otherwise you might get hundreds of thousands of values and that would make the autocompletion meaningless.
Will Grafana UI allow you to insert an abritrary client id if you simply return an empty list here?

Reply:
That's a great question, and I believe we also briefly discussed this with @max-vogler. I'll research it.

Reply:
Ok so I have some findings.

There are [different types of Grafana variables](https://grafana.com/docs/grafana/latest/variables/templates-and-variables/#variable-types). A user may add a Grafana variable of type [text box](https://grafana.com/docs/grafana/latest/variables/variable-types/add-text-box-variable/#add-a-text-box-variable) (instead of the default [query type](https://grafana.com/docs/grafana/latest/variables/variable-types/add-query-variable/#add-a-query-variable) variable), which will *not* issue a request by Grafana for a list of possible values. Within this text box variable, a user can type in the relevant Client ID, all of that without support for auto-completion of course.

In this scenario, we might decide to not support the second type of response of the `/search`, endpoint which is issued when the default query type is selected for the Grafana variable. This is also discussed in this thread https://github.com/google/grr/pull/832#discussion_r477532744.

What do you think @mbushkov?

Reply:
Following an offline discussion with @mbushkov, I'll remove the method `fetchClientIds` from the Grrafana object and the function `GetClientIdsFromFleetspeak` from `fleetspeak_utils.py` that support this functionality.